parser2/src/tokenize.rs

```diff
@@ -0,0 +1,585 @@
+// use std::fmt;
+use std::str::CharIndices;
+use unicode_xid::UnicodeXID;
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Error {
+    EmptyCharLiteral,
+    UnderIndented,
+    UnexpectedToken,
+    UnexpectedEscapeCode(char),
+    UnterminatedCharLiteral,
+    UnterminatedStringLiteral,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Token<'input> {
+    // Keywords
+    And,
+    Else,
+    If,
+    In,
+    Let,
+    Match,
+    Then,
+    Type,
+    With,
+
+    Name(&'input str),
+    Operator(&'input str),
+    ByteLiteral(u8),
+    IntLiteral(i64),
+    FloatLiteral(f64),
+    StringLiteral(String),
```

Comment:
Might be a performance problem if `StringLiteral` gets cloned a lot and there is a large literal (not sure how LALRPOP handles it). The old parser does the same but since `&'input str` is already used maybe that should be used here as well? The processing of escapes could be done later on then.